File: practice-coding-test/java-kotlin/src/kr/co/programmers/java/monthly_code_challenge/season1/CountingAfterQuadCompression.java
package kr.co.programmers.java.monthly_code_challenge.season1;

import java.util.Arrays;

public class CountingAfterQuadCompression {

    public static void main(String[] args) {
//        int[][] arr = {
//                {1, 1, 0, 0},
//                {1, 0, 0, 0},
//                {1, 0, 0, 1},
//                {1, 1, 1, 1}};
        int[][] arr = {
                {1, 1, 1, 1, 1, 1, 1, 1},
                {0, 1, 1, 1, 1, 1, 1, 1},
                {0, 0, 0, 0, 1, 1, 1, 1},
                {0, 1, 0, 0, 1, 1, 1, 1},
                {0, 0, 0, 0, 0, 0, 1, 1},
                {0, 0, 0, 0, 0, 0, 0, 1},
                {0, 0, 0, 0, 1, 0, 0, 1},
                {0, 0, 0, 0, 1, 1, 1, 1}
        };
        System.out.println(Arrays.toString(new Solution().solution(arr)));
    }

    static class Solution {

        public int[] solution(int[][] arr) {
            int[] answer = {0, 0};
            checkQuad(arr, answer, 0, 0, arr.length);
            return answer;
        }

        private void checkQuad(int[][] arr, int[] count, int x, int y, int length) {
            if (length == 1) {
                count[arr[x][y]]++;
                return;
            }

            int sum = 0;
            for (int i = x; i < x + length; i++) {
                for (int j = y; j < y + length; j++) {
                    sum += arr[i][j];
                }
            }

            if (sum == 0) {
                count[0]++;
            } else if (sum == length * length) {
                count[1]++;
            } else {
                int newLength = length / 2;
                checkQuad(arr, count, x, y, newLength);
                checkQuad(arr, count, x + newLength, y, newLength);
                checkQuad(arr, count, x, y + newLength, newLength);
                checkQuad(arr, count, x + newLength, y + newLength, newLength);
            }
        }
    }
}

File: practice-coding-test/javascript/src/monthly-code-challenge/season1/counting-after-quad-compression.js
// @ts-check

/**
 * @param {number[][]} arr
 * @param {number[]} answer
 * @param {number} x
 * @param {number} y
 * @param {number} length
 */
function countingAfterQuadCompression(arr, answer, x, y, length) {
  if (length === 1) {
    answer[arr[x][y]] += 1
    return
  }

  let sum = 0
  for (let i = x; i < x + length; i += 1) {
    for (let j = y; j < y + length; j += 1) {
      sum += arr[i][j]
    }
  }

  if (sum === 0) {
    answer[0] += 1
  } else if (sum === length * length) {
    answer[1] += 1
  } else {
    const newLength = Math.floor(length / 2)
    countingAfterQuadCompression(arr, answer, x, y, newLength)
    countingAfterQuadCompression(arr, answer, x + newLength, y, newLength)
    countingAfterQuadCompression(arr, answer, x, y + newLength, newLength)
    countingAfterQuadCompression(
      arr,
      answer,
      x + newLength,
      y + newLength,
      newLength
    )
  }
}

/**
 * @param {number[][]} arr
 * @returns {number[]}
 */
function solution(arr) {
  const answer = [0, 0]
  countingAfterQuadCompression(arr, answer, 0, 0, arr.length)
  return answer
}

/**
 * @param {number[][]} arr
 * @returns {number[]}
 */
function diffSolution(arr) {
  const quadZip = (row, col, n) => {
    if (n < 2) {
      return arr[row][col] ? [0, 1] : [1, 0]
    }
    let cnt0 = 0
    let cnt1 = 0
    n >>= 1

    for (let [i, j] of [
      [0, 0],
      [0, 1],
      [1, 0],
      [1, 1],
    ]) {
      let [zero, one] = quadZip(row + i * n, col + j * n, n)
      cnt0 += zero
      cnt1 += one
    }

    if (cnt0 === 0) return [0, 1]
    if (cnt1 === 0) return [1, 0]
    return [cnt0, cnt1]
  }
  return quadZip(0, 0, arr.length)
}

const arr = [
  [1, 1, 1, 1, 1, 1, 1, 1],
  [0, 1, 1, 1, 1, 1, 1, 1],
  [0, 0, 0, 0, 1, 1, 1, 1],
  [0, 1, 0, 0, 1, 1, 1, 1],
  [0, 0, 0, 0, 0, 0, 1, 1],
  [0, 0, 0, 0, 0, 0, 0, 1],
  [0, 0, 0, 0, 1, 0, 0, 1],
  [0, 0, 0, 0, 1, 1, 1, 1],
]

console.log(solution(arr))
console.log(diffSolution(arr))
